vpr/src/backend/physical_netlist_writer.cpp

```diff
@@ -0,0 +1,732 @@
+#include "physical_netlist_writer.h"
+
+#include <algorithm>
+#include <fstream>
+#include <iomanip>
+#include <iostream>
+#include <iterator>
+#include <kj/std/iostream.h>
+#include <set>
+#include <sstream>
+#include <string>
+#include <unordered_map>
+#include <zlib.h>
+
+#include "globals.h"
+
+#include "rr_metadata.h"
+
+#include "vtr_assert.h"
+#include "vtr_logic.h"
+#include "vtr_version.h"
+#include "vtr_util.h"
+#include "vpr_error.h"
+
+#include "atom_netlist_utils.h"
+#include "netlist_writer.h"
+#include "vpr_utils.h"
+#include "route_tree_timing.h"
+
+static void write_message(::capnp::MallocMessageBuilder& message, const std::string& filename) {
+    kj::Array<capnp::word> words = messageToFlatArray(message);
+    kj::ArrayPtr<kj::byte> bytes = words.asBytes();
+
+    gzFile file = gzopen(filename.c_str(), "w");
+    VTR_ASSERT(file != Z_NULL);
+
+    VTR_ASSERT(gzwrite(file, &bytes[0], bytes.size()) == (int)bytes.size());
+    VTR_ASSERT(gzclose(file) == Z_OK);
+}
+
+void write_physical_netlist(t_vpr_setup& vpr_setup) {
+    auto& device_ctx = g_vpr_ctx.device();
+
+    // Opening device file
+    gzFile file = gzopen(vpr_setup.FileNameOpts.ArchFile.c_str(), "r");
+    VTR_ASSERT(file != Z_NULL);
+
+    std::vector<uint8_t> output_data;
+    output_data.resize(4096);
+    std::stringstream sstream(std::ios_base::in | std::ios_base::out | std::ios_base::binary);
+    while (true) {
+        int ret = gzread(file, output_data.data(), output_data.size());
+        VTR_ASSERT(ret >= 0);
+        if (ret > 0) {
+            sstream.write((const char*)output_data.data(), ret);
+            VTR_ASSERT(sstream);
+        } else {
+            VTR_ASSERT(ret == 0);
+            int error;
+            gzerror(file, &error);
+            VTR_ASSERT(error == Z_OK);
+            break;
+        }
+    }
+
+    VTR_ASSERT(gzclose(file) == Z_OK);
+
+    sstream.seekg(0);
+    kj::std::StdInputStream istream(sstream);
+
+    // Reader options
+    capnp::ReaderOptions reader_options;
+    reader_options.nestingLimit = std::numeric_limits<int>::max();
+    reader_options.traversalLimitInWords = std::numeric_limits<uint64_t>::max();
+
+    capnp::InputStreamMessageReader message_reader(istream, reader_options);
+
+    auto device_reader = message_reader.getRoot<DeviceResources::Device>();
+
+    ::capnp::MallocMessageBuilder message;
+    PhysicalNetlist::PhysNetlist::Builder physical_netlist = message.initRoot<PhysicalNetlist::PhysNetlist>();
+    physical_netlist.setPart(vpr_setup.device_layout);
+
+    ic::InterchangeWriterVisitor visitor(&device_ctx.arch->strings, device_reader, physical_netlist);
+    NetlistWalker nl_walker(visitor);
+    nl_walker.walk();
+
+    std::string circuit_name = vtr::split(vpr_setup.FileNameOpts.CircuitFile, ".")[0];
+    std::string phys_netlist_filename = circuit_name + ".phys";
```

Comment:
Can we have a flag for specifying the output name explicitly?

Reply:
I'd keep it as is, to follow also the way other files are produced. We can add a flag later on